### PR TITLE
Add basic tests for sync/pagination with vector clock tokens.

### PR DESCRIPTION
**Pull request by erikjohnston**
These are tests for #8439

### CODE REVIEW

File: synapse/handlers/room.py [hunk 1]
```diff
@@ -576,6 +577,10 @@ async def create_room(
                 values to go in the body of the 'join' event (typically
                 `avatar_url` and/or `displayname`.
 
+            requested_room_id: Allow callees to request a particular room ID.
+                This is useful for testing event persistence sharding, and
+                *should not* be exposed to clients.
+
```
Comment (richvdh) [inline synapse/handlers/room.py:583]: this feels like a bit of a footgun waiting to be misused. Can you call `store_room` and `_send_events_for_new_room` from the test?

Comment (richvdh) [reply synapse/handlers/room.py:583]: alternatively: patch out `_generate_room_id`.

File: tests/replication/test_sharded_event_persister.py [hunk 2]
```diff
@@ -100,3 +103,171 @@ def test_basic(self):
...
+        user_id = self.register_user("user", "pass")
+        access_token = self.login("user", "pass")
+        requester = create_requester(user_id, access_token)
+
+        store = self.hs.get_datastore()
+
+        # Create two room on the different workers.
+        room_creator = self.hs.get_room_creation_handler()
+        self.get_success(
+            room_creator.create_room(requester, {}, requested_room_id=room_id1), by=0.1
```
Comment (richvdh) [inline tests/replication/test_sharded_event_persister.py:144]: I'm surprised that the `by` params are needed here

**Review by richvdh — CHANGES_REQUESTED** [review]
is it worth testing that we can paginate forwards from a vector clock token?

File: tests/replication/test_sharded_event_persister.py (hunk 2 above)
Comment (erikjohnston) [reply tests/replication/test_sharded_event_persister.py:144]: I'm not sure why, but we seem to need it whenever we send HTTP requests (e.g. the `render` methods automatically advance the clock repeatedly until a result comes in), which happens here where it sends HTTP replication requests to send events.

**Review by richvdh — APPROVED** [review]
lgtm

**Pull request merged**